File: api/ai/agents/providers/openai_agent.py
from typing import List

from openai import OpenAI
from openai.types.chat import ChatCompletion

from ai.agents.generic_agent import GenericAgent
from constants import MODEL_GPT
from models import MessageDto


class OpenAiAgent(GenericAgent):

    def __init__(self, name):
        self.client: OpenAI = OpenAI()
        self.name = name
        self.model = MODEL_GPT

    def ask(self, chat_messages: List[MessageDto], is_json: bool = False) -> str | None:
        self.logger.debug(f"Asking {self.name} agent. Message history for this player: {chat_messages[-1].msg}")
        for msg in chat_messages[1:]:  # Skip the first message because it's a long instruction
            self.logger.debug(f"{msg.role}: {msg.msg}")

        chat_completion: ChatCompletion = self.client.chat.completions.create(
            messages=[{"role": msg.role.value, "content": msg.msg} for msg in chat_messages],
            model=self.model,

            # Controls randomness: lowering results in less random completions.
            # As the temperature approaches zero, the model will become deterministic
            # and repetitive.
            temperature=0.5,

            # If set, partial message deltas will be sent.
            stream=False,
            response_format={"type": "json_object"} if is_json else {"type": "text"},
        )
        resp = chat_completion.choices[0].message.content
        self.logger.info(f"{self.name}: {resp}")
        return resp

    def ask_wth_text(self, question: str, is_json: bool = False) -> str | None:
        self.logger.debug(f"Asking {self.name} agent: {question}")
        chat_completion: ChatCompletion = self.client.chat.completions.create(
            messages=[{"role": "user", "content": question}],
            model=self.model,

            # Controls randomness: lowering results in less random completions.
            # As the temperature approaches zero, the model will become deterministic
            # and repetitive.
            temperature=0.5,

            # If set, partial message deltas will be sent.
            stream=False,
            response_format={"type": "json_object"} if is_json else {"type": "text"},
        )
        resp = chat_completion.choices[0].message.content
        return resp

File: werewolf-client/app/ai/OpenAIAgent.ts
import {AbstractAgent} from "@/app/ai/AbstractAgent";
import OpenAI from "openai";
import {User} from "@/app/api/models";
import {AgentMessageDto, LLMModel} from "@/app/ai/models";

export class OpenAIAgent extends AbstractAgent {
    constructor(name: string, user: User) {
        const client = new OpenAI({
            apiKey: user.keys[LLMModel.GPT4].value,
        })
        super("rand", name, 'TBD', LLMModel.GPT4, 0.2, client);
    }

    async ask(messages: AgentMessageDto[]): Promise<string | null> {
        this.logger(`Asking agent. Message history: ${messages[messages.length - 1].msg}`);

        const cl = this.client as OpenAI

        try {
            const response = await this.client.createChatCompletion({
                model: this.model,
                messages: messages.map(msg => ({
                    role: msg.role,
                    content: msg.msg
                })),
                temperature: this.temperature,
            });

            const reply = response.data.choices[0].message?.content;
            this.logger(`Reply: ${reply}`);
            return reply || null;
        } catch (error) {
            console.error('Error in OpenAIAgent.ask:', error);
            return null;
        }
    }
}